scAML_embedding
================

# Introduction

To get some first experience with multimodal single cell data, we use
the data set generated by (Granja et al. 2019). In this studies, authors
generated multimodal data including combined scRNA- and surface epitopes
(CITE-seq) and scATAC seq data from healthy donor peripheral blood
(PBMCs) and bone marrow mononuclear cells (BMMCs) and CD34+ enriched
BMMCs. CITE-seq data was integrated and used to generate a reference map
for healthy hematopoiesis. Furthermore, they used scATAC-seq to generate
a reference epigenetic map of healthy hematopoiesis.

They tested the robustness and generalizability of the reference map by
iteratively projection various data sets of healthy bone marrow
subcompartements onto the map always confirming high agreement in
subtype projection and identification.

Building on this multimodal reference map they then profile PB/BMMCs
from 6 distinct patients with mixed-phenotype acute leukemia (MPAL),
project this data set into the reference map and investigate
associations to non-malignent cellular counterparts, differentially
expressed peaks, TFs and compared them to other published single-cell
leukemia data sets.

With this vignette we’ll try to

1.  load the scRNA- and scATAC-seq data from the healthy samples
2.  create an seurat object and perform basic quality controls,
    filtering etc.
3.  perform a multimodal integration of scRNA and scATAC-seq
4.  project the MPAL data into the WNN-graph
5.  use IReNA to construct GRNs

For steps 1-4, since they are well standardized, we’ll work with the
seurat vignette.

# Results

## scRNA-seq

### Loading, Filtering

The GEO data set includes scRNA-seq data BMMCs, CD34+ BMMCs, PBCMs and 6
MPAL samples of which two are first diagnosis and relapse from the same
patient.

scRNA-seq data are stored as a sparse matrix

``` r
BMMC1.scRNA <- readRDS("datasets/granja_scMPAL/GSM4138872_scRNA_BMMC_D1T1.rds")
BMMC2.scRNA <- readRDS("datasets/granja_scMPAL/GSM4138873_scRNA_BMMC_D1T2.rds")
```

Both objects include 20287 rows (=transcripts) from \~ 6300 single cells
(identified by a common prefix (in this case “BMMC_D1T2:”) and a cell
UMI. The first goal would be to get this data into an seurat object.

``` r
str(BMMC1.scRNA)
```

    Lade nötiges Paket: Matrix

    Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
      ..@ i       : int [1:9089992] 34 42 50 59 73 74 100 104 112 113 ...
      ..@ p       : int [1:6271] 0 1352 3353 4794 6603 7709 9089 10082 11498 12775 ...
      ..@ Dim     : int [1:2] 20287 6270
      ..@ Dimnames:List of 2
      .. ..$ : chr [1:20287] "FAM138A" "OR4F5" "AL627309.1" "OR4F29" ...
      .. ..$ : chr [1:6270] "BMMC_D1T1:AAACCCAAGATGCAGC-1" "BMMC_D1T1:AAACCCACAAACTCGT-1" "BMMC_D1T1:AAACCCACAGTGTACT-1" "BMMC_D1T1:AAACCCATCGCTATTT-1" ...
      ..@ x       : num [1:9089992] 1 1 1 1 1 1 1 1 1 1 ...
      ..@ factors : list()

SeuratObject includes a CreateSeuratObject wrapper which takes 10x
CellRanger input. Merging two 10x runs into a single seurat object is
described
[here](https://satijalab.org/seurat/articles/merge_vignette.html).

``` r
library(Seurat)
```

    Attaching SeuratObject

``` r
library(dplyr)
```


    Attache Paket: 'dplyr'

    Die folgenden Objekte sind maskiert von 'package:stats':

        filter, lag

    Die folgenden Objekte sind maskiert von 'package:base':

        intersect, setdiff, setequal, union

``` r
BMMC1.scRNA.seurat <- CreateSeuratObject(counts = BMMC1.scRNA,project = "BMMC1.scRNA", assay = "RNA", names.field = 2, names.delim = ":")

BMMC2.scRNA.seurat <- CreateSeuratObject(counts = BMMC2.scRNA,project = "BMMC2.scRNA", assay = "RNA", names.field = 2, names.delim = ":")

BMMC.scRNA.seurat <- merge(BMMC1.scRNA.seurat, y = BMMC2.scRNA.seurat, add.cell.ids = c("T1", "T2"), project = "BMMC.scRNA")
```

Lets have a first glimpse into out newly generated seurat object. We can
access the assays data by using \[\[“assay”\]\].

``` r
glimpse(head(BMMC.scRNA.seurat[["RNA"]]@counts, 5))
```

    Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
      ..@ i       : int [1:681] 2 2 2 2 2 2 2 2 2 2 ...
      ..@ p       : int [1:12603] 0 0 0 0 0 0 0 0 0 0 ...
      ..@ Dim     : int [1:2] 5 12602
      ..@ Dimnames:List of 2
      .. ..$ : chr [1:5] "FAM138A" "OR4F5" "AL627309.1" "OR4F29" ...
      .. ..$ : chr [1:12602] "T1_BMMC_D1T1:AAACCCAAGATGCAGC-1" "T1_BMMC_D1T1:AAACCCACAAACTCGT-1" "T1_BMMC_D1T1:AAACCCACAGTGTACT-1" "T1_BMMC_D1T1:AAACCCATCGCTATTT-1" ...
      ..@ x       : num [1:681] 1 2 3 1 1 2 2 1 1 2 ...
      ..@ factors : list()

Some cell-wise meta data is automatically calculated during the creation
of the seurat object.

``` r
head(BMMC.scRNA.seurat@meta.data, 5)
```

                                     orig.ident nCount_RNA nFeature_RNA
    T1_BMMC_D1T1:AAACCCAAGATGCAGC-1 BMMC1.scRNA       2433         1352
    T1_BMMC_D1T1:AAACCCACAAACTCGT-1 BMMC1.scRNA       5106         2001
    T1_BMMC_D1T1:AAACCCACAGTGTACT-1 BMMC1.scRNA       3589         1441
    T1_BMMC_D1T1:AAACCCATCGCTATTT-1 BMMC1.scRNA       3603         1809
    T1_BMMC_D1T1:AAACGAACACCCAATA-1 BMMC1.scRNA       2065         1106

Basic metrics include number of total counts per cell (nCount_RNA) and
number of non-zero features (nFeature_RNA) per cell. This helps to
identify cells with very low (\<200) or very high (\>25.000) unique
features, the former being dead and the latter duplicated cells. In
principal, we could identify cells with high (\>5%) mitochondrial counts
which is also a feature of low-quality samples by adding a new meta data
column. However, mitochondrial and ribosomal genes were already filtered
out by the authors.

``` r
rownames(BMMC.scRNA.seurat) %>% grepl(pattern = "^MT-") %>% any()
```

    [1] FALSE

``` r
BMMC.scRNA.seurat[["percent.mt"]] <- PercentageFeatureSet(BMMC.scRNA.seurat, pattern = "^MT-")

VlnPlot(BMMC.scRNA.seurat, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"))
```

    Warning in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents =
    idents, : All cells have the same value of percent.mt.

![](scAML_seurat_files/figure-gfm/unnamed-chunk-7-1.png)

``` r
FeatureScatter(BMMC.scRNA.seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

![](scAML_seurat_files/figure-gfm/unnamed-chunk-8-1.png)

The authors set the QC filtering threshold to remove cells with less
than 400 detected genes (features) and less than 1000 counts (UMIs) as
well as above 10.000 counts (UMIs). As shown by the violin plot, this
filtering was already applied to the data set.

``` r
BMMC.scRNA.seurat <- subset(BMMC.scRNA.seurat, nFeature_RNA > 400 & nCount_RNA > 1000 & nCount_RNA < 10000)
```

### Normalization

The default normalization of scaling to 10.000 depth followed by log2
(n+1) normalization was also used by the authors in the manuscript.

``` r
BMMC.scRNA.seurat <- NormalizeData(BMMC.scRNA.seurat, normalization.method = "LogNormalize", scale.factor = 1000)
```

### Variable Feature Detection and PCA

We detect the top 3000 most variable features (genes) to later use as
input for dimensionality reduction and visualization

``` r
BMMC.scRNA.seurat <- FindVariableFeatures(BMMC.scRNA.seurat, selection.method = "vst", nfeatures = 3000)
```

``` r
VariableFeatures(BMMC.scRNA.seurat) %>% head(20)
```

     [1] "HBB"    "HBA2"   "HBA1"   "HBD"    "HBM"    "AHSP"   "CA1"    "PRTN3" 
     [9] "ELANE"  "CTSG"   "PTGDS"  "GNLY"   "AZU1"   "GYPA"   "DNTT"   "MPO"   
    [17] "ALAS2"  "CLC"    "GYPB"   "FCER1A"

We next scale feature-wise to make sure that gene expression is not the
main driver of variance. We scale all features instead of only the 3000
most variable ones.

``` r
BMMC.scRNA.seurat <- ScaleData(BMMC.scRNA.seurat, features = rownames(BMMC.scRNA.seurat))
```

    Centering and scaling data matrix

The scaled data matrix can be accessed

``` r
BMMC.scRNA.seurat[["RNA"]]@scale.data[1:10, 1:10]
```

               T1_BMMC_D1T1:AAACCCAAGATGCAGC-1 T1_BMMC_D1T1:AAACCCACAAACTCGT-1
    FAM138A                         0.00000000                      0.00000000
    OR4F5                           0.00000000                      0.00000000
    AL627309.1                     -0.21944680                     -0.21944680
    OR4F29                          0.00000000                      0.00000000
    OR4F16                          0.00000000                      0.00000000
    FAM87B                         -0.01379727                     -0.01379727
    LINC00115                      -0.09746042                     -0.09746042
    FAM41C                         -0.17345448                     -0.17345448
    AL645608.2                      0.00000000                      0.00000000
    SAMD11                         -0.03336301                     -0.03336301
               T1_BMMC_D1T1:AAACCCACAGTGTACT-1 T1_BMMC_D1T1:AAACCCATCGCTATTT-1
    FAM138A                         0.00000000                      0.00000000
    OR4F5                           0.00000000                      0.00000000
    AL627309.1                     -0.21944680                     -0.21944680
    OR4F29                          0.00000000                      0.00000000
    OR4F16                          0.00000000                      0.00000000
    FAM87B                         -0.01379727                     -0.01379727
    LINC00115                      -0.09746042                     -0.09746042
    FAM41C                         -0.17345448                     -0.17345448
    AL645608.2                      0.00000000                      0.00000000
    SAMD11                         -0.03336301                     -0.03336301
               T1_BMMC_D1T1:AAACGAACACCCAATA-1 T1_BMMC_D1T1:AAACGAACAGCAGTCC-1
    FAM138A                         0.00000000                      0.00000000
    OR4F5                           0.00000000                      0.00000000
    AL627309.1                     -0.21944680                     -0.21944680
    OR4F29                          0.00000000                      0.00000000
    OR4F16                          0.00000000                      0.00000000
    FAM87B                         -0.01379727                     -0.01379727
    LINC00115                      -0.09746042                     -0.09746042
    FAM41C                         -0.17345448                     -0.17345448
    AL645608.2                      0.00000000                      0.00000000
    SAMD11                         -0.03336301                     -0.03336301
               T1_BMMC_D1T1:AAACGAACATGACGAG-1 T1_BMMC_D1T1:AAACGAATCAAGCCAT-1
    FAM138A                         0.00000000                      0.00000000
    OR4F5                           0.00000000                      0.00000000
    AL627309.1                     -0.21944680                     -0.21944680
    OR4F29                          0.00000000                      0.00000000
    OR4F16                          0.00000000                      0.00000000
    FAM87B                         -0.01379727                     -0.01379727
    LINC00115                      -0.09746042                     -0.09746042
    FAM41C                         -0.17345448                     -0.17345448
    AL645608.2                      0.00000000                      0.00000000
    SAMD11                         -0.03336301                     -0.03336301
               T1_BMMC_D1T1:AAACGCTAGCCGTAAG-1 T1_BMMC_D1T1:AAACGCTAGCTAAACA-1
    FAM138A                         0.00000000                      0.00000000
    OR4F5                           0.00000000                      0.00000000
    AL627309.1                     -0.21944680                      3.11097437
    OR4F29                          0.00000000                      0.00000000
    OR4F16                          0.00000000                      0.00000000
    FAM87B                         -0.01379727                     -0.01379727
    LINC00115                      -0.09746042                     -0.09746042
    FAM41C                         -0.17345448                     -0.17345448
    AL645608.2                      0.00000000                      0.00000000
    SAMD11                         -0.03336301                     -0.03336301

``` r
BMMC.scRNA.seurat <- RunPCA(BMMC.scRNA.seurat, features = VariableFeatures(BMMC.scRNA.seurat))
```

    PC_ 1 
    Positive:  S100A9, LYZ, CST3, S100A8, FCN1, TYROBP, MNDA, VCAN, S100A11, CTSS 
           GRN, FTL, S100A6, CD14, PSAP, FCER1G, S100A12, S100A4, TYMP, MS4A6A 
           CSTA, CYBB, LST1, PLAUR, SERPINA1, TSPO, CTSD, LGALS1, APLP2, CEBPD 
    Negative:  NPM1, EEF1B2, LDHB, PTMA, LTB, CD3E, HSPA8, CD3D, HINT1, GYPC 
           HNRNPA1, CD69, HMGN1, CD81, IL7R, UBB, CD7, SEPT6, IL32, PEBP1 
           CD3G, TCF7, ISG20, LEF1, CD27, HSP90AB1, PIK3IP1, SNHG7, SPOCK2, PSIP1 
    PC_ 2 
    Positive:  JUNB, CD3E, CD3D, LTB, IL32, CD7, CD3G, IL7R, ARL4C, TCF7 
           TNFAIP3, CD27, SPOCK2, PIK3IP1, CD2, CD247, CD6, IFITM1, BCL11B, GZMM 
           FLT3LG, TRABD2A, RORA, ZAP70, CCR7, DUSP2, MAL, GIMAP7, OXNAD1, CD69 
    Negative:  STMN1, TUBA1B, TUBB, TYMS, HMGB1, IGLL1, H2AFZ, H2AFY, HMGN2, HMGA1 
           FABP5, SOX4, NREP, CENPU, MCM7, CDT1, PCNA, CDCA7, MYB, PRSS57 
           YBX3, YWHAE, CKS2, ZWINT, NUSAP1, HIST1H4C, BIRC5, PPIA, HMGB3, RRM2 
    PC_ 3 
    Positive:  SPIB, CD79B, CD74, CD79A, HLA-DRA, FAM129C, TCL1A, IGLL5, VPREB3, HLA-DPA1 
           HLA-DQB1, FCRLA, HLA-DPB1, MZB1, TCF4, IRF4, HLA-DQA1, BCL7A, HLA-DRB5, HLA-DMA 
           AFF3, CD72, HLA-DRB1, CD19, BLK, HLA-DMB, CXXC5, CYB561A3, TSPAN13, ABHD15 
    Negative:  PRSS57, GAPDH, ENO1, LDHB, SERPINB1, MPO, HSPB1, CLEC11A, AZU1, ANXA1 
           CYTL1, SRGN, C1QTNF4, ELANE, LAPTM4B, SPINK2, CTSG, EMID1, GYPC, MS4A3 
           PRTN3, HSPD1, CD3E, NPW, HSP90AB1, EGFL7, TPI1, ITM2A, NPR3, ANP32B 
    PC_ 4 
    Positive:  SPINK2, EGFL7, CYTL1, LAPTM4B, CD34, LILRA4, NPR3, ITM2C, IL3RA, PRSS57 
           PPP1R14B, EMID1, CRHBP, HSP90AB1, BCL11A, AVP, C1QTNF4, MSRB3, IRF8, SERPINF1 
           IL18, APP, BAALC, PROM1, TPM2, SMPD3, TSC22D1, CDK6, IRF7, MDK 
    Negative:  TOP2A, NUSAP1, MKI67, RRM2, UBE2C, BIRC5, CDK1, HMGB2, AURKB, CENPF 
           TPX2, HIST1H4C, KIFC1, CCNA2, CDKN3, CCNB2, TUBB, ASPM, DLGAP5, PTTG1 
           ZWINT, TYMS, GTSE1, HMMR, CDCA5, CDCA8, NCAPG, NUF2, TUBB4B, H2AFX 
    PC_ 5 
    Positive:  LILRA4, GZMB, IL3RA, C12orf75, SMPD3, CLEC4C, PLD4, SERPINF1, IRF7, IRF8 
           SCT, PTPRS, CCDC50, UGCG, CLIC3, TPM2, PTCRA, GAS6, SLC15A4, LILRB4 
           SHD, ZFAT, PACSIN1, MAPKAPK2, APP, DERL3, ITM2C, CIB2, LGMN, PLAC8 
    Negative:  YBX3, CD79B, CD79A, VPREB3, IGLL5, RCSD1, CD9, TCL1A, NEIL1, SSBP2 
           PCDH9, MME, CD19, EBF1, EGFL7, DTX1, VPREB1, LHPP, CD34, CD72 
           PRDX1, HIST1H2AC, POU2AF1, MS4A1, SPINK2, PAX5, CYTL1, BCL7A, CD22, LAT2 

``` r
DimPlot(BMMC.scRNA.seurat, reduction = "pca")
```

![](scAML_seurat_files/figure-gfm/unnamed-chunk-16-1.png)

### Determining Data Set Dimensionality

``` r
BMMC.scRNA.seurat <- JackStraw(BMMC.scRNA.seurat, num.replicate = 100, dims = 50)
```

``` r
BMMC.scRNA.seurat <- ScoreJackStraw(BMMC.scRNA.seurat, dims = 1:50)
```

``` r
JackStrawPlot(BMMC.scRNA.seurat, dims = 1:50)
```

    Warning: Removed 110009 rows containing missing values (geom_point).

![](scAML_seurat_files/figure-gfm/unnamed-chunk-19-1.png)

From the jackstraw plot it appears that PCs 1:45 are the most
informative.

``` r
ElbowPlot(BMMC.scRNA.seurat, ndims = 50)
```

![](scAML_seurat_files/figure-gfm/unnamed-chunk-20-1.png)

### Visualization

First, we identify clusters/communities based on the KNN graph.

``` r
BMMC.scRNA.seurat <- FindNeighbors(BMMC.scRNA.seurat, dims = 1:45)
```

    Computing nearest neighbor graph

    Computing SNN

``` r
BMMC.scRNA.seurat <- FindClusters(BMMC.scRNA.seurat, resolution = 0.8)
```

    Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

    Number of nodes: 12602
    Number of edges: 547260

    Running Louvain algorithm...
    Maximum modularity in 10 random starts: 0.9063
    Number of communities: 23
    Elapsed time: 1 seconds

And visualize the data set using UMAP

``` r
BMMC.scRNA.seurat <- RunUMAP(BMMC.scRNA.seurat, dims = 1:50, n.neighbors = 35, min.dist = 0.45)
```

    Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
    To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
    This message will be shown once per session

    20:17:18 UMAP embedding parameters a = 0.67 b = 1.278

    20:17:18 Read 12602 rows and found 50 numeric columns

    20:17:18 Using Annoy for neighbor search, n_neighbors = 35

    20:17:18 Building Annoy index with metric = cosine, n_trees = 50

    0%   10   20   30   40   50   60   70   80   90   100%

    [----|----|----|----|----|----|----|----|----|----|

    **************************************************|
    20:17:20 Writing NN index file to temp file C:\Users\wolf-\AppData\Local\Temp\Rtmpq4z8pX\file4b74246a46f0
    20:17:20 Searching Annoy index using 1 thread, search_k = 3500
    20:17:23 Annoy recall = 100%
    20:17:24 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 35
    20:17:25 Initializing from normalized Laplacian + noise (using irlba)
    20:17:26 Commencing optimization for 200 epochs, with 642972 positive edges
    20:17:38 Optimization finished

``` r
DimPlot(BMMC.scRNA.seurat, reduction = "umap")
```

![](scAML_seurat_files/figure-gfm/unnamed-chunk-22-1.png)

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-granja2019" class="csl-entry">

Granja, Jeffrey M., Sandy Klemm, Lisa M. McGinnis, Arwa S. Kathiria,
Anja Mezger, M. Ryan Corces, Benjamin Parks, et al. 2019. “Single-Cell
Multiomic Analysis Identifies Regulatory Programs in Mixed-Phenotype
Acute Leukemia.” *Nature Biotechnology* 37 (12): 1458–65.
<https://doi.org/10.1038/s41587-019-0332-7>.

</div>

</div>
